ostreedev/ostree
# Alternative signing system

Proposed the implementation of alternative signing system (as followup of https://github.com/ostreedev/ostree/issues/1233)

Approach is pretty simple: have a common interface for signing and verification, with implementation of particular sign/verify in separate modules.
- could co-exists with current GPG implementation
- allows to store and use multiple signature types in detached metadata in the same time
- `--with-libsodium` option is needed to build with support of `ed25515` signing engine 

### Current status:
- added the very first version of signing interface
- added "dummy" signing engine
- added "ed25519" signing engine
- added new builtin `ostree sign` (inspired by "ostree gpg-sign") allowing to sign and verify commits
- able to sign/verify commits
- able to sign/verify summary file
- added tests for commits and summary signing/verification
- added tests for pulling verification

not implemented:
- have no support of printing the status on the screen (yet)
- probably not integrated into all places there it should be used (need hints here)

### New configuration keys
For repository and remotes:
- `sign-verify` -- global and per-remote to trigger signature verification of updates
- `verification-key` -- per-remote -- for `ed25519`: base64 encoded public key to use for verification
 - `verification-file` -- per-remote -- for `ed25519`: file with the list of base64 public keys to  use for verification

### Dummy engine
Accept any ASCII string as public/secret key. Used mostly for testing the signing interface itself. Support only single public key for verification.

### Ed25519
Added "well-known" system places for `ed25519` public keys -- expected 1 base64 key per line:
- files
  - `/etc/ostree/trusted.ed25519`
  - `DATADIR + /ostree/trusted.ed25519`
- and directories:
  - `/etc/ostree/trusted.ed25519.d`
  - `DATADIR + /ostree/trusted.ed25519.d`

The same is for revoked keys:
  - `/etc/ostree/revoked.ed25519`
  - `/etc/ostree/revoked.ed25519.d`
  - `DATADIR + /ostree/revoked.ed25519`
  - `DATADIR + /ostree/rvokeded.ed25519.d`

Current logic for verification during the commits/summary file pulling:
- use `verification-key` if it exists in configuration
- use `verification-file` if it exists in configuration
- check public keys from well-known system places
